File: compare-sort-algorithms/src/bubble_sort.js
/**
 * Bubble Sort implementation in JavaScript with performance measurement.
 */

const fs = require('fs');
const path = require('path');

/**
 * Bubble Sort implementation.
 * Repeatedly steps through the list, compares adjacent elements, and swaps them if they are in the wrong order.
 * 
 * @param {number[]} arr - The array to be sorted
 */
function bubbleSort(arr) {
    const n = arr.length;
    
    for (let i = 0; i < n; i++) {
        // Flag to optimize if no swaps are made in a pass
        let swapped = false;
        
        // Last i elements are already in place
        for (let j = 0; j < n - i - 1; j++) {
            // Compare adjacent elements
            if (arr[j] > arr[j + 1]) {
                // Swap arr[j] and arr[j+1]
                [arr[j], arr[j + 1]] = [arr[j + 1], arr[j]];
                swapped = true;
            }
        }
        
        // If no swapping occurred in this pass, array is sorted
        if (!swapped) {
            break;
        }
    }
}

/**
 * Reads numbers from file and returns as array.
 * 
 * @param {string} filename - The file to read from
 * @returns {number[]} Array of numbers
 */
function readNumbersFromFile(filename) {
    try {
        const data = fs.readFileSync(filename, 'utf8');
        return data.trim().split('\n').map(line => parseInt(line.trim(), 10)).filter(num => !isNaN(num));
    } catch (err) {
        console.error(`Error: ${filename} not found. Please run generate_data.py first.`);
        process.exit(1);
    }
}

/**
 * Verifies if the array is sorted.
 * 
 * @param {number[]} arr - The array to check
 * @returns {boolean} True if sorted, false otherwise
 */
function isSorted(arr) {
    for (let i = 0; i < arr.length - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}

// Main function
function main() {
    console.log("JavaScript Bubble Sort Performance Test");
    console.log("=======================================");
    
    // Parse command line arguments
    const args = process.argv.slice(2);
    let filename = "random_list.txt";
    let resultFilename = "results_javascript.txt";
    
    if (args.length > 0) {
        filename = args[0];
        console.log(`Using dataset: ${filename}`);
    }
    
    if (args.length > 1) {
        resultFilename = args[1];
    }
    
    // Read data from file
    console.log("Reading data from file...");
    const data = readNumbersFromFile(filename);
    console.log(`Data size: ${data.length} integers`);
    
    // Create a copy for sorting (to preserve original)
    const dataCopy = [...data];
    
    // Measure sorting time
    console.log("Starting Bubble Sort...");
    const startTime = process.hrtime();
    bubbleSort(dataCopy);
    const endTime = process.hrtime(startTime);
    
    // Calculate execution time in seconds
    const executionTime = endTime[0] + endTime[1] / 1e9;
    
    // Verify the array is sorted
    const sorted = isSorted(dataCopy);
    
    // Results
    console.log(`Sorting completed: ${sorted ? 'SUCCESS' : 'FAILED'}`);
    console.log(`Execution time: ${executionTime.toFixed(6)} seconds`);
    console.log(`Elements per second: ${Math.floor(data.length / executionTime)}`);
    
    // Save results to file
    const results = `JavaScript Bubble Sort Results\n` +
                    `Data size: ${data.length}\n` +
                    `Execution time: ${executionTime.toFixed(6)} seconds\n` +
                    `Elements per second: ${Math.floor(data.length / executionTime)}\n` +
                    `Sorted correctly: ${sorted}\n`;
    
    fs.writeFileSync(resultFilename, results);
    console.log(`Results saved to ${resultFilename}`);
}

// Run the main function
main();

File: compare-sort-algorithms/src/bubble_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <stdbool.h>
#include <string.h>

/**
 * Bubble Sort implementation.
 * Repeatedly steps through the list, compares adjacent elements, and swaps them if they are in the wrong order.
 */
void bubbleSort(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        // Flag to optimize if no swaps are made in a pass
        bool swapped = false;
        
        // Last i elements are already in place
        for (int j = 0; j < n - i - 1; j++) {
            // Compare adjacent elements
            if (arr[j] > arr[j + 1]) {
                // Swap arr[j] and arr[j+1]
                int temp = arr[j];
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                swapped = true;
            }
        }
        
        // If no swapping occurred in this pass, array is sorted
        if (!swapped) {
            break;
        }
    }
}

/**
 * Reads numbers from file and returns as array.
 */
int* readNumbersFromFile(const char* filename, int* size) {
    FILE* file = fopen(filename, "r");
    if (file == NULL) {
        printf("Error: %s not found. Please run generate_data.py first.\n", filename);
        exit(1);
    }
    
    // Count the number of lines in the file
    int count = 0;
    char buffer[256];
    while (fgets(buffer, sizeof(buffer), file) != NULL) {
        count++;
    }
    
    // Allocate memory for the array
    int* numbers = (int*)malloc(count * sizeof(int));
    if (numbers == NULL) {
        printf("Memory allocation failed.\n");
        fclose(file);
        exit(1);
    }
    
    // Reset file pointer and read the numbers
    rewind(file);
    int i = 0;
    while (fgets(buffer, sizeof(buffer), file) != NULL) {
        numbers[i++] = atoi(buffer);
    }
    
    fclose(file);
    *size = count;
    return numbers;
}

/**
 * Verifies if the array is sorted.
 */
bool isSorted(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}

int main(int argc, char* argv[]) {
    printf("C Bubble Sort Performance Test\n");
    printf("===============================\n");
    
    // Parse command line arguments
    const char* filename = "random_list.txt";
    const char* resultFilename = "results_c.txt";
    
    if (argc > 1) {
        filename = argv[1];
        printf("Using dataset: %s\n", filename);
    }
    
    if (argc > 2) {
        resultFilename = argv[2];
    }
    
    // Read data from file
    printf("Reading data from file...\n");
    int size;
    int* data = readNumbersFromFile(filename, &size);
    printf("Data size: %d integers\n", size);
    
    // Create a copy for sorting (to preserve original)
    int* dataCopy = (int*)malloc(size * sizeof(int));
    if (dataCopy == NULL) {
        printf("Memory allocation failed.\n");
        free(data);
        exit(1);
    }
    memcpy(dataCopy, data, size * sizeof(int));
    
    // Measure sorting time
    printf("Starting Bubble Sort...\n");
    clock_t start = clock();
    bubbleSort(dataCopy, size);
    clock_t end = clock();
    
    double executionTime = ((double)(end - start)) / CLOCKS_PER_SEC;
    
    // Verify the array is sorted
    bool sorted = isSorted(dataCopy, size);
    
    // Results
    printf("Sorting completed: %s\n", sorted ? "SUCCESS" : "FAILED");
    printf("Execution time: %.6f seconds\n", executionTime);
    printf("Elements per second: %.0f\n", size / executionTime);
    
    // Save results to file
    FILE* resultFile = fopen(resultFilename, "w");
    if (resultFile == NULL) {
        printf("Error: Could not open %s for writing.\n", resultFilename);
        free(data);
        free(dataCopy);
        exit(1);
    }
    
    fprintf(resultFile, "C Bubble Sort Results\n");
    fprintf(resultFile, "Data size: %d\n", size);
    fprintf(resultFile, "Execution time: %.6f seconds\n", executionTime);
    fprintf(resultFile, "Elements per second: %.0f\n", size / executionTime);
    fprintf(resultFile, "Sorted correctly: %s\n", sorted ? "true" : "false");
    
    fclose(resultFile);
    printf("Results saved to %s\n", resultFilename);
    
    // Cleanup
    free(data);
    free(dataCopy);
    
    return 0;
}
